### PR TITLE
fix(subscriptions): Duplicate charge on cvc fail

### PR DESCRIPTION
**Pull request by StaberindeZA**
## Because

- An incorrect card cvc isn't always immediately caught, which causes an
  incomplete subscription. Once the customer corrects the cvc it could
  happen that a new subscription is created, instead of completing the
  existing subscription. As a result, both subscriptions are completed
  and the card is charged twice.

## This pull request

- Ensures that only 1 subscription is created and the card is only
  charged once.

## Issue that this pull request solves

Closes: #11024

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [x] I have added necessary documentation (if appropriate).
- [x] I have verified that my changes render correctly in RTL (if appropriate).

## Screenshots (Optional)

Please attach the screenshots of the changes made in case of change in user interface.

## Other information (Optional)

Any other information that is important to this pull request.

### CODE REVIEW

**Comment by StaberindeZA** [comment]
TLDR - In Production, when a customer enters an incorrect card CVC, receives and error message, **refreshes** the page, enters the correct information, and successfully completes the payment, they will have 2 active subscriptions, and their card will have been charged twice.

### Why does this happen?
1. When the payment method is attached to the customer, the CVC check is "unchecked", and successfully attached.
1. When the subscription is created, the CVC check has been completed by Stripe, and is now "fail", however the subscription is still created.
1. We show an error message, and ask the user to try again.
1.1. If the user clicks the "Try again" button, everything works as expected our "Retry" process is followed, and the initial invoice should be paid, if the user enters correct payment information.
1.1. If instead the user refreshes the page, or navigates away and comes back later, a second subscription is created, and both subscriptions are charged.

### Next steps options.
1. Delete the incomplete subscription
1. Update our payment process, so that if a customer has an incomplete subscription, the Retry process is followed instead of a new subscription being created.
1. Let Support handle these cases.
1.1. It looks like this only really happens when the customer enters an incorrect CVC and then refreshes the page instead of hitting Retry.

### Notes
- I haven't been able to replicate this on DEV/Stage. The CVC cards fail on payment method attach, not on subscription create.
- If you'd like to see an example of this in Prod, ping me on Slack and I can share a customer.

File: packages/fxa-payments-server/src/lib/stripe.ts [hunk 1]
```diff
@@ -249,6 +252,12 @@ export async function handlePaymentIntent({
         paymentMethodId
       ) {
         apiDetachFailedPaymentMethod({ paymentMethodId });
+        // Delete incomplete subscription.
+        // If we do not do this, and a new subscription is added, with valid payment card,
+        // then both subscriptions will be paid, resulting in a duplicate charge.
+        // A new subscription will be created if the customer refreshes the page,
+        // instead of hitting the try again button.
+        apiDeleteSubscriptionId({ subscriptionId });
```
Comment (chenba) [inline packages/fxa-payments-server/src/lib/stripe.ts:260]: Where is this `subscriptionId` from?  Can we delete on the back end without having the front end trigger it?

Comment (StaberindeZA) [reply packages/fxa-payments-server/src/lib/stripe.ts:260]: Yup definitely. Doing it in the backend would probably be even better.

File: packages/fxa-auth-server/lib/payments/stripe.ts [hunk 2]
```diff
@@ -489,6 +489,20 @@ export class StripeHelper {
       },
       { idempotencyKey: `ssc-${subIdempotencyKey}` }
     );
+
+    const paymentIntent = (subscription.latest_invoice as Stripe.Invoice)
```
Comment (StaberindeZA) [inline packages/fxa-auth-server/lib/payments/stripe.ts:493]: @chenba what do you think. Something like this? (I'll clean it up still)

Comment (chenba) [reply packages/fxa-auth-server/lib/payments/stripe.ts:493]: I don't actually know whether Stripe has already attempted a charge at this point.  If it did, then this probably should work.

Comment (StaberindeZA) [reply packages/fxa-auth-server/lib/payments/stripe.ts:493]: From the few examples that I've seen and from what the discord chat, it does look like a charge has been attempted.

**Comment by StaberindeZA** [comment]
I tested this locally by replacing the output of the [stripe.subscription.create](https://github.com/mozilla/fxa/blob/c98e2b3cf0c0b7b55377cab80946b399244339c6/packages/fxa-auth-server/lib/payments/stripe.ts#L482) call with the testing fixture added as part of this PR.

Other than that, I haven't been able to find a way to test this in Local or Stage.

**Comment by chenba** [comment]
> The CVC cards fail on payment method attach, not on subscription create.

So doesn't that mean we know about the failure _before_ we create the subscription?  Why not prevent the subscription creation then?  When the request to create a subscription has a payment method id we already load the payment method (see `createSubscriptionWithPMI`), is this something we can check there?

**Comment by StaberindeZA** [comment]
> nit: "CVV" I think?
> 
> But also suggested an alternative approach to look into.

Looks like Visa uses CVV (Card Verification Value) and Master Card uses CVC (Card Verification Code). TIL

In Stripe at least it seems they use CVC everywhere I've seen, so I'd like to keep it matched with what they have.

**Comment by StaberindeZA** [comment]
> > The CVC cards fail on payment method attach, not on subscription create.
> 
> So doesn't that mean we know about the failure _before_ we create the subscription? Why not prevent the subscription creation then? When the request to create a subscription has a payment method id we already load the payment method (see `createSubscriptionWithPMI`), is this something we can check there?

That would work on the behaviour that we see in Stage, but in Production, it doesn't always return the error on Payment Method attach. This PR is meant to specifically fix the scenario where the CVC check only fails on subscription create.

I did mention it in the [long comment](https://github.com/mozilla/fxa/pull/12046#issuecomment-1054457762) under heading "Why does this happen?", but it might be easier to see it. I can share a customer with you where this happened.

**Comment by StaberindeZA** [comment]
> > The CVC cards fail on payment method attach, not on subscription create.
> 
> So doesn't that mean we know about the failure _before_ we create the subscription? Why not prevent the subscription creation then? When the request to create a subscription has a payment method id we already load the payment method (see `createSubscriptionWithPMI`), is this something we can check there?

At the moment before the subscription creation we don't know about the failure. The Payment Method Attach returns `cvc_check: 'unchecked'`. 

I was going to suggest that, if `cvc_check === 'unchecked'`, then do `GET /v1/payment_methods/:id`, until `cvc_check` comes back as `pass` or `fail`, and only on a `pass` continue to create the subscription.

But then I came across these 2 pages, that made me think the `cvc_check` will only change to `fail` after the card is charged. It looks like declines after payment are things to be expected in Stripe, so now I'm not sure if that approach would work.
1. [Check if a card is valid without a charge](https://support.stripe.com/questions/check-if-a-card-is-valid-without-a-charge)
1. [Understanding declines and failed payments](https://stripe.com/docs/declines)

I haven't read through everything in those docs yet, but I did also find [this conversation in the Stripe Discord](https://discord.com/channels/841573134531821608/841573134531821616/905923208279097385), where the Stripe dev seemed to suggest you won't know about the failure before the subscription is created.

**Comment by chenba** [comment]
> 
> Looks like Visa uses CVV (Card Verification Value) and Master Card uses CVC (Card Verification Code).
> 
 
haha of course.

**Comment by chenba** [comment]
> [snip]

Thanks for the thorough response.  I'll review the current approach now.

File: packages/fxa-auth-server/test/local/payments/fixtures/stripe/ subscription_pmi_expanded_incomplete_cvc_fail.json [hunk 3]
```diff
@@ -0,0 +1,498 @@
+{
```
Comment (chenba) [inline packages/fxa-auth-server/test/local/payments/fixtures/stripe/ subscription_pmi_expanded_incomplete_cvc_fail.json:1]: Is that a space as the first character of the filename?

**Review by chenba — APPROVED** [review]
LGTM.  Thanks again for the detailed explanation.
